### PR TITLE
Issue #678 and #677

### PR DESCRIPTION
**Pull request by lovemaths**
Moved TheoryData below test case, and changed the method in WsFederationMetadataSerializer.

### CODE REVIEW

File: src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs [hunk 1]
```diff
@@ -141,50 +147,46 @@ protected virtual void ReadKeyDescriptorForSigning(WsFederationConfiguration con
...
             {
                 if (reader.IsStartElement(Elements.KeyDescriptor, Namespaces.MetadataNamespace) && reader.GetAttribute(Attributes.Use).Equals(keyUse.Signing))
-                    ReadKeyDescriptorForSigning(configuration, reader);
+                    roleDescriptor.KeyInfos.Add(ReadKeyDescriptorForSigning(reader));
                 else if (reader.IsStartElement(Elements.SecurityTokenEndpoint, Namespaces.FederationNamespace))
-                    ReadSecurityTokenEndpoint(configuration, reader);
+                    roleDescriptor.TokenEndpoint = ReadSecurityTokenEndpoint(reader);
                 else if (reader.IsStartElement())
                     reader.ReadOuterXml();
                 else
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs]: LogMessages.IDX13003 doesn't describe this exception completely. 
It should more descriptive.

File: src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs [hunk 2]
```diff
@@ -141,50 +147,46 @@ protected virtual void ReadKeyDescriptorForSigning(WsFederationConfiguration con
...
 
+            var roleDescriptor = new SecurityTokenServiceTypeRoleDescriptor();
+
             // <RoleDescriptorr>
             reader.ReadStartElement();
             while (reader.IsStartElement())
             {
                 if (reader.IsStartElement(Elements.KeyDescriptor, Namespaces.MetadataNamespace) && reader.GetAttribute(Attributes.Use).Equals(keyUse.Signing))
-                    ReadKeyDescriptorForSigning(configuration, reader);
+                    roleDescriptor.KeyInfos.Add(ReadKeyDescriptorForSigning(reader));
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs]: We need tests for each of these conditions.

File: src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs [hunk 3]
```diff
@@ -269,4 +272,28 @@ internal static string Trim(string stringToTrim)
             return stringToTrim.Trim(charsToTrim);
         }
     }
+
+    /// <summary>
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs]: Put public classes into a seperate file.

File: src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs [hunk 4]
```diff
@@ -269,4 +272,28 @@ internal static string Trim(string stringToTrim)
...
+        } = new List<KeyInfo>();
+
+        /// <summary>
+        /// Token endpoint
+        /// </summary>
+        public string TokenEndpoint
+        {
+            get;
+            set;
+        } = null;
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs]: no need to assign null.

File: src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs (hunk 2 above)
Comment (lovemaths) [inline src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs]: Already covered by the existing tests.

**Comment by lovemaths** [comment]
@brentschmaltz  rebased.

**Comment by brentschmaltz** [comment]
:shipit:

**Pull request merged**